library/pkcs11_client.c

```diff
@@ -0,0 +1,527 @@
+        {CKA_EXTRACTABLE, &ck_extractable, sizeof( ck_extractable )},
+#define COMMON_PRIVATE_ATTRIBUTES 7 // number of attributes above
+        {-1, NULL, 0},
+        {-1, NULL, 0},
+        {-1, NULL, 0},
+        {-1, NULL, 0},
+        {-1, NULL, 0},
+        {-1, NULL, 0},
+        {-1, NULL, 0},
+        {-1, NULL, 0},
+    };
+    CK_ATTRIBUTE *public_end = public_attributes + COMMON_PUBLIC_ATTRIBUTES;
+    CK_ATTRIBUTE *private_end = private_attributes + COMMON_PRIVATE_ATTRIBUTES;
+#undef COMMON_PUBLIC_ATTRIBUTES
+#undef COMMON_PRIVATE_ATTRIBUTES
+    unsigned char *data = NULL;
+    CK_RV rv;
+
+    if( hPublicKey != NULL )
+        *hPublicKey = CK_INVALID_HANDLE;
+    if( hPrivateKey != NULL )
+        *hPrivateKey = CK_INVALID_HANDLE;
+
+    /* Prepare the data-dependent key attributes */
+    switch( mbedtls_pk_representation_type( mbedtls_pk_get_type( ctx ) ) )
+    {
+#if defined(MBEDTLS_ECDSA_C)
+        case MBEDTLS_PK_ECKEY:
+        {
+            const mbedtls_ecp_keypair *ec = mbedtls_pk_ec( *ctx );
+            unsigned char *p;
+            size_t curve_bytes = ( ec->grp.pbits + 7 ) / 8;
+            size_t point_bytes = 4 + 2 * curve_bytes; // overapproximation
+            int format = MBEDTLS_ECP_PF_UNCOMPRESSED;
+            int ret;
+            data = mbedtls_calloc( 1,
+                                   MBEDTLS_OID_EC_GRP_MAX_SIZE +
+                                   ( hPublicKey == NULL ? 0 : point_bytes ) +
+                                   ( hPrivateKey == NULL ? 0 : curve_bytes ) );
+            if( data == NULL )
+            {
+                rv = CKR_HOST_MEMORY;
+                goto exit;
+            }
+            p = data;
+            ck_key_type = CKK_ECDSA;
+            /* Convert the group identifier */
+            ret = mbedtls_ecp_ansi_write_group( &ec->grp, p,
+                                                MBEDTLS_OID_EC_GRP_MAX_SIZE );
+            if( ret < 0 )
+            {
+                rv = CKR_GENERAL_ERROR;
+                goto exit;
+            }
+            public_end->type = CKA_EC_PARAMS;
+            public_end->pValue = p;
+            public_end->ulValueLen = ret;
+            p += ret;
+            *private_end++ = *public_end++;
+            if( hPublicKey != NULL )
+            {
+                /* Convert the public point */
+                ret = mbedtls_ecp_ansi_write_point( ec, format, p, point_bytes );
+                if( ret < 0 )
+                {
+                    rv = CKR_GENERAL_ERROR;
+                    goto exit;
+                }
+                public_end->type = CKA_EC_POINT;
+                public_end->pValue = p;
+                public_end->ulValueLen = ret;
+                p += ret;
+                public_end++;
+            }
+            if( hPrivateKey != NULL )
+            {
+                /* Convert the private value */
+                MPI_TO_CK( &ec->d, private_end++, CKA_VALUE, &p, curve_bytes );
```

Comment:
Style: I think the increment would be more visible on a line on its own.

Reply:
I'm not sure if that's what you mean, but would you like to have the "private_end++" in its own line, separately? It seems like a bit of an overkill for me.